language/proto/gen/update_proto_csv.go

```diff
@@ -66,25 +72,225 @@ func main() {
 	flag.Parse()
 
 	if *protoCsvPath == "" {
-		log.Fatal("-proto must be set")
-	}
-	if *googleapisRootPath == "" {
-		log.Fatal("-go_googleapis must be set")
+		log.Fatal("-proto_csv must be set")
 	}
 
 	protoContent := &bytes.Buffer{}
 	protoContent.WriteString(prefix)
 
-	err := filepath.Walk(*googleapisRootPath, func(path string, info os.FileInfo, err error) error {
+	if *goGoogleapisRootPath != "" {
+		if err := generateFromPath(*goGoogleapisRootPath, protoContent); err != nil {
+			log.Fatal(err)
+		}
+	} else if *comGoogleGoogleapisRootPath != "" {
+		if err := generateFromQuery(*bazelPath, *comGoogleGoogleapisRootPath, protoContent); err != nil {
+			log.Fatal(err)
+		}
+	} else {
+		log.Fatal("either -com_google_googleapis or -go_googleapis must be set")
+	}
+
+	if err := ioutil.WriteFile(*protoCsvPath, protoContent.Bytes(), 0666); err != nil {
+		log.Fatal(err)
+	}
+}
+
+//
+// Process -com_google_googleapis case
+//
+const repoPrefix = "@com_google_googleapis"
+
+var bazelQuery = []string{
+	`query`,
+	`kind("proto_library rule", //google/...)`,
+	`--output`,
+	`xml`,
+	`--experimental_enable_repo_mapping`,
+}
+
+type Query struct {
+	XMLName xml.Name `xml:"query"`
+	Version string   `xml:"version,attr"`
+	Rules   []Rule   `xml:"rule"`
+}
+
+type Rule struct {
+	XMLName xml.Name `xml:"rule"`
+	Class   string   `xml:"class,attr"`
+	Name    string   `xml:"name,attr"`
+	Strings []String `xml:"string"`
+	Labels  []Label  `xml:"label"`
+	Lists   []List   `xml:"list"`
+}
+
+type List struct {
+	XMLName xml.Name `xml:"list"`
+	Name    string   `xml:"name,attr"`
+	Labels  []Label  `xml:"label"`
+}
+
+type Label struct {
+	XMLName xml.Name `xml:"label"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+type String struct {
+	XMLName xml.Name `xml:"string"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+func generateFromQuery(bazelPath, workspacePath string, protoContent io.Writer) error {
```

Comment:
Nit: rename `protoContent` to `w`, and make it the first parameter.

Reply:
Changed (and also changed the `generateFromPath` signature for consistency). In general, the naming for many of the params/variables duplicate the original version (i.e. the `generateByPath` version.